subsys/net/lib/lwm2m/lwm2m_obj_software_mgmt.c

```diff
@@ -0,0 +1,611 @@
+			"no more room: %u", obj_inst_id);
+		return NULL;
+	}
+
+	/* Set default values */
+	update_state[index] = SW_MGMT_UPDATE_STATE_INITIAL;
+	activation_state[index] = SW_MGMT_ACTIVATION_STATE_DISABLED;
+	update_result[index] = SW_MGMT_UPDATE_RESULT_DEFAULT;
+	package_uri[index][0] = '\0';
+	package_name[index][0] = '\0';
+	package_version[index][0] = '\0';
+
+	(void)memset(res[index], 0,
+			 sizeof(res[index][0]) * ARRAY_SIZE(res[index]));
+	init_res_instance(res_inst[index], ARRAY_SIZE(res_inst[index]));
+
+	/* initialize instance resource data */
+	INIT_OBJ_RES_DATA(SOFTWARE_MGMT_PACKAGE_NAME_ID, res[index], i,
+			  res_inst[index], j,
+			  &package_name[index], PACKAGE_NAME_LEN);
+	INIT_OBJ_RES_DATA(SOFTWARE_MGMT_PACKAGE_VERSION_ID, res[index], i,
+			  res_inst[index], j,
+			  &package_version[index], PACKAGE_VERSION_LEN);
+	INIT_OBJ_RES_OPT(SOFTWARE_MGMT_PACKAGE_ID, res[index], i, res_inst[index], j, 1, false,
+			 true, NULL, NULL, NULL, package_write_cb, NULL);
+	INIT_OBJ_RES(SOFTWARE_MGMT_PACKAGE_URI_ID, res[index], i, res_inst[index], j, 1, false,
+			 true, package_uri[index], PACKAGE_URI_LEN,
+			 NULL, NULL, NULL, package_uri_write_cb, NULL);
+	INIT_OBJ_RES_EXECUTE(SOFTWARE_MGMT_INSTALL_ID, res[index], i, software_mgmt_install_cb);
+	INIT_OBJ_RES_EXECUTE(SOFTWARE_MGMT_UNINSTALL_ID, res[index], i, software_mgmt_uninstall_cb);
+	INIT_OBJ_RES_DATA(SOFTWARE_MGMT_UPDATE_STATE_ID, res[index], i, res_inst[index], j,
+			  &update_state[index], sizeof(update_state[index]));
+	INIT_OBJ_RES_DATA(SOFTWARE_MGMT_UPDATE_RESULT_ID, res[index], i, res_inst[index], j,
+			  &update_result[index], sizeof(update_result[index]));
+	INIT_OBJ_RES_EXECUTE(SOFTWARE_MGMT_ACTIVATE_ID, res[index], i, software_mgmt_activate_cb);
+	INIT_OBJ_RES_EXECUTE(SOFTWARE_MGMT_DEACTIVATE_ID, res[index], i,
+			     software_mgmt_deactivate_cb);
+	INIT_OBJ_RES_DATA(SOFTWARE_MGMT_ACTIVATION_STATE_ID, res[index], i, res_inst[index], j,
+			  &activation_state[index], sizeof(activation_state[index]));
+
+	inst[index].resources = res[index];
+	inst[index].resource_count = i;
+	LOG_DBG("Create LWM2M Software Management instance: %d", obj_inst_id);
+	return &inst[index];
+}
+
+static int lwm2m_software_mgmt_init(const struct device *dev)
+{
+	struct lwm2m_engine_obj_inst *obj_inst = NULL;
+	int ret = 0;
+
+	software_mgmt.obj_id = LWM2M_OBJECT_SOFTWARE_MANAGEMENT_ID;
+	software_mgmt.version_major = SOFTWARE_MGMT_VERSION_MAJOR;
+	software_mgmt.version_minor = SOFTWARE_MGMT_VERSION_MINOR;
+	software_mgmt.is_core = true;
+	software_mgmt.fields = fields;
+	software_mgmt.field_count = ARRAY_SIZE(fields);
+	software_mgmt.max_instance_count = MAX_INSTANCE_COUNT;
+	software_mgmt.create_cb = software_mgmt_create;
+	lwm2m_register_obj(&software_mgmt);
+
+	/* auto create the first instance */
+	ret = lwm2m_create_obj_inst(LWM2M_OBJECT_SOFTWARE_MANAGEMENT_ID, 0, &obj_inst);
+	if (ret < 0) {
+		LOG_ERR("Create LWM2M server instance 0 error: %d", ret);
+	}
```

Comment:
I just wonder, does it make sense to create an instance by default? Perhps it'd be better to leave it up to the user to define what software/firmware components can be updated?